Fix 500 error when attempting to update CommCareUser user_data via API

gherceg opened the pull request:
## Product Description
<!-- For non-invisible changes, describe user-facing effects. -->

## Technical Summary
<!--
    Provide a link to the ticket or document which prompted this change,
    Describe the rationale and design decisions.
-->
https://dimagi-dev.atlassian.net/browse/SAAS-12580

Pretty straightforward bug unless I'm missing context on why we wouldn't want to allow updating user metadata individually. My assumption is there was just a mistake and `should_save = True` was not included at the end of the `elif key == 'user_data':` block. Fixed that and added tests that would have caught this bug, as well as some of the other fields being updated individually. ~~It felt potentially excessive to have a isolated test for every mutable field of a user to ensure it saves properly, but I'm not opposed to adding that in if others feel it is worth it.~~ Took Matt's suggestion and moved tests to the unit level for the `_update` method.
## Feature Flag
<!-- If this is specific to a feature flag, which one? -->

## Safety Assurance

### Safety story
<!--
Describe how you became confident in this change, such as
local testing, why the change is inherently safe, and/or plans to limit the blast radius of a defect.
-->
This one field was not behaving like the rest due to a missing line so now that it is aligned with the rest of the fields, it feels safe.
### Automated test coverage

<!-- Identify the related test coverage and the tests it would catch -->
Added a test that would have caught this bug, along with tests that would catch similar issues with other fields.
### QA Plan

<!--
- Describe QA plan that along with automated test coverages proves this PR is regression free
- Link to QA Ticket
-->
No QA necessary.

### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

gherceg committed:
set should_save to True for user_data if no exception raised

mjriley reviewed (approved):
The change looks good. I think the tests are probably fine, but these seem like integration tests to me, where they're high enough level that is fairly hard to know what is being tested. I think it might have been clearer to write tests against `obj_update` or even the `_update` method, to avoid the noise with constructing API requests. Especially if it actually was the case that user data was only meant to be updated with group changes, that test would be a lot easier to read at the `obj_update` level.

gherceg committed:
Merge branch 'master' into gh/bugs/user-api/user-data-500

gherceg committed:
add unit tests for _update method; remove redundant integration level tests

gherceg committed:
return list of errors instead of should_save

gherceg committed:
run isort

gherceg commented:
I'd review commit by commit, as there were hardly any changes to the `v0_5.py` file, but the diff makes it looks like. I just moved the `_update` method to bottom of the class, made it a class method to make it easier to test, replaced `should_save` with `errors` that were appended to and returned, and use the validation error exception `e.message` instead of `str(e)`. 

e6737028c8de50f682a783ca19762f44782ad31c is taking Matt's suggestion of adding unit tests for the `_update` method instead of integration level tests, and 62bdee68a006d37cd76eb1e6b1cb6b786117c8f2 is returning `errors` from `_update` instead of `should_save`, since it should save if no errors are encountered. I suppose I am missing a test case for an empty body, in which case it would still save. Maybe the better behavior is to return an error in that case.

gherceg commented:
FYI @esoergel and @snopoke since this is in the same area as https://github.com/dimagi/commcare-hq/pull/31345 which you just reviewed.

gherceg committed:
add integration test for multiple errors; caught bug

gherceg committed:
unused import

gherceg committed:
add unit test for groups

snopoke commented on corehq/apps/api/tests/user_resources.py:
```diff
@@ -35,6 +38,8 @@
 
 from .utils import APIResourceTest
+from ..resources.v0_5 import CommCareUserResource
+from ...domain.shortcuts import create_domain
```
I'm not sure if this is best practice by I think relative imports should be restricted to the same app.

snopoke reviewed (approved):
One comment but not blocking

gherceg committed:
change relative import from outside of app
